### PR TITLE
LEAF-2381 - Email No recipient

### PR DESCRIPTION
**Pull request by shaneodd**
Add in logic to make sure it does not attempt to send email if there are no recipients set for sending the email. (Line 303)

Additional work was setup with coding standards so more than a single line of change.

### CODE REVIEW

**Review by maxrichardjr — APPROVED** [review]
Did the legacy guys not specify type on a lot of things. Dangerous!'

**Review by raphaelfretter — APPROVED** [review]

**Review by pete-nerantzinis — APPROVED** [review]

**Review by Pelentan — APPROVED** [review]

**Review by aerinkayne — DISMISSED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by AmandaSumner — DISMISSED** [review]

File: LEAF_Request_Portal/Email.php [hunk 1]
```diff
@@ -100,7 +102,7 @@ function getFilepath($tpl, $type = '')
     /**
      * Removes all email addresses from object recipient variable
      */
-    public function clearRecipients()
+    public function clearRecipients(): void
     {
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:106]: Shouldn't the PHPDoc contain the return type even if it's void?

File: LEAF_Request_Portal/Email.php [hunk 2]
```diff
@@ -109,16 +111,16 @@ public function clearRecipients()
...
         $this->emailSender = $strAddress;
     }
 
     /**
      * Clean and Set subject of email object variable
-     * @param $strSubject
+     * @param string $strSubject
      */
-    public function setSubject($strSubject)
+    public function setSubject(string $strSubject): void
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:123]: PHPDoc no @return

File: LEAF_Request_Portal/Email.php [hunk 3]
```diff
@@ -127,12 +129,13 @@ public function setSubject($strSubject)
...
-     * @param $strContent   - content to add to template
-     * @param $tplVar       - variable within template
-     * @param $tplFile      = template file name
+     * @param string $objVar       - private variable in Email object
+     * @param string $strContent   - content to add to template
+     * @param string $tplVar       - variable within template
+     * @param string $tplFile      = template file name
      */
-    public function setContent($tplFile, $tplVar = '', $strContent = '') {
+    public function setContent(string $tplFile, string $tplVar = '', string $strContent = ''): string
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:137]: PHPDoc no @return

File: LEAF_Request_Portal/Email.php [hunk 4]
```diff
@@ -178,22 +182,19 @@ public function emailActiveNotAlreadyAdded($address) {
...
     {
-        if (preg_match('/(\w+@[a-zA-Z_)+?\.[a-zA-Z]{2,6})/', $address) == 0)
-        {
+        if (preg_match('/(\w+@[a-zA-Z_)+?\.[a-zA-Z]{2,6})/', $address) == 0){
             return false;
         }
 
-        if ($this->emailRecipient == '')
-        {
+        if ($this->emailRecipient == ''){
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:195]: PSR12 requires a space between ) and {

File: LEAF_Request_Portal/Email.php [hunk 5]
```diff
@@ -205,27 +206,26 @@ public function addRecipient($address, $requiredAddress = false)
 
     /**
      * Adds all users in a given Position to the receipient object variable list
-     * @param $positionID
+     * @param int $positionID
      */
-    public function addPositionRecipient($positionID)
+    public function addPositionRecipient(int $positionID): void
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:211]: PHPDoc no @return

File: LEAF_Request_Portal/Email.php [hunk 6]
```diff
@@ -205,27 +206,26 @@ public function addRecipient($address, $requiredAddress = false)
...
         }
     }
 
     /**
      * Adds all users in a given Group to the reeeipient object variable list
-     * @param $groupID
+     * @param int $groupID
      */
-    public function addGroupRecipient($groupID)
+    public function addGroupRecipient(int $groupID): void
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:228]: PHPDoc no @return I won't add anymore if there are more for this

File: LEAF_Request_Portal/Email.php [hunk 7]
```diff
@@ -296,60 +294,65 @@ public function sendMail()
...
         $emailQueueName = sha1($emailCache . random_int(0, 99999999));
-        if (strlen(trim($emailCache)) == 0)
-        {
+        if (strlen(trim($emailCache)) == 0) {
             trigger_error('Mail error: ' . $this->emailSubject);
+            return false;
+        }
 
+        // if we have no recipients then we should not create the email.
+        if(strlen($email['recipient']) == 0) {
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:303]: PSR12 - if ( space is missing

File: LEAF_Request_Portal/Email.php [hunk 8]
```diff
@@ -490,10 +490,11 @@ function setTemplateByLabel($emailTemplateLabel)
     /**
      * Given email template location where email addresses are stored
      * Get the email addresses, line by line, and add them if valid to CC or BCC
-     * @param $tplLocation
+     * @param string $tplLocation
+     * @param bool $isCc
      * @param false $isBcc
      */
-    function setEmailToCcWithTemplate($tplLocation, $isCc = false)
+    function setEmailToCcWithTemplate(string $tplLocation, bool $isCc = false): void
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:497]: Does not appear as though $isBcc is used anywhere. It is not defined within the method definition and I didn't see it being used locally. probably shouldn't be in the PHPDoc either

File: LEAF_Request_Portal/Email.php [hunk 9]
```diff
@@ -527,10 +528,10 @@ function setSubjectWithTemplate($subjectTemplate)
     /**
      * Purpose (deprecated): set email body directly from passed in HTML
      * LEGACY: Included as scripts created by portal uses that implement sends using this feature
-     * @param $i
+     * @param string $i
      * @throws SmartyException
      */
-    public function setBody($i)
+    public function setBody(string $i): void
     {
```
Comment (jampaul3) [inline LEAF_Request_Portal/Email.php:535]: I don't see where the SmartyException is used unless it's within the $smarty->fetch, in which case are we supposed to show that in the PHPDoc of this method?

**Review by jampaul3 — APPROVED** [review]

**Comment by jampaul3** [comment]
I added several comments to this PR, none of which should prevent it being pushed to prod, Just felt that if we are making changes to a file to this significance then we ought to do it complete and correct to PSR12 standard.

**Review by jrodcode — DISMISSED** [review]

**Pull request merged**